### PR TITLE
feat: switch to driver writing files

### PR DESCRIPTION
**Pull request by aramase**
Signed-off-by: Anish Ramasekar <anish.ramasekar@gmail.com>

**Reason for Change**:

- Adds an optional feature gated with a flag `--driver-write-secrets` to allow the CSI driver to write secrets (this feature is available in CSI Driver **v0.0.21+**). Default is false. We will switch to true in the next releases based on testing.
- Templates the e2e kind test pipeline 
- Removed the custom cloud env test pipeline as it wasn't testing anything. Instead added e2e tests for custom cloud environment with kind.
- Adds test pipeline for with and without `driver-write-secrets` feature.

**Requirements**

- [x] squashed commits
- [x] included documentation
- [x] added unit tests and e2e tests (if applicable).

**Issue Fixed**:

fixes https://github.com/Azure/secrets-store-csi-driver-provider-azure/issues/475
fixes https://github.com/kubernetes-sigs/secrets-store-csi-driver/issues/460

**Does this change contain code from or inspired by another project?**

- [ ] Yes
- [x] No

**If "Yes," did you notify that project's maintainers and provide attribution?**

**Special Notes for Reviewers**:

### CODE REVIEW

File: pkg/provider/provider.go [hunk 1]
```diff
@@ -283,15 +285,22 @@ func (p *Provider) MountSecretsStoreObjectContent(ctx context.Context, attrib ma
...
+			klog.InfoS("successfully wrote file", "file", fileName, "pod", klog.ObjectRef{Namespace: p.PodNamespace, Name: p.PodName})
+		} else {
+			// these files will returned to the CSI driver as part of grpc response
+			files[fileName] = objectContent
 		}
-		klog.InfoS("successfully wrote file", "file", fileName, "pod", klog.ObjectRef{Namespace: p.PodNamespace, Name: p.PodName})
 	}
 
-	return objectVersionMap, nil
+	return files, objectVersionMap, nil
```
[the remark this answers is not in the document]
Comment (aramase) [reply pkg/provider/provider.go:303]: The interesting part is at this point, we don't really know if the user has configured a larger size using the `--max-call-recv-msg-size`. The driver will log the error which will `ResourceExhausted`. We can enhance the driver logs to log setting a larger size when that's the error but a warning in provider might not be required (as it's a driver configuration). WDYT?

[the remark this answers is not in the document]
Comment (nilekhc) [reply pkg/provider/provider.go:303]: `ResourceExhausted` logs in driver looks good. Should we create an issue to enhance driver ogs?

[the remark this answers is not in the document]
Comment (aramase) [reply pkg/provider/provider.go:303]: Opened https://github.com/kubernetes-sigs/secrets-store-csi-driver/issues/536 to track it

**Review by nilekhc — APPROVED** [review]
lgtm

File: .pipelines/templates/e2e-test-kind.yaml [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+            AZURE_CLIENT_SECRET: $(AZURE_CLIENT_SECRET)
+            KEY_NAME: $(KEY_NAME)
+            KEY_VERSION: $(KEY_VERSION)
+            KEYVAULT_NAME: $(KEYVAULT_NAME)
+            SECRET_NAME: $(SECRET_NAME)
+            TENANT_ID: $(TENANT_ID)
+            CI_KIND_CLUSTER: true
+            AZURE_ENVIRONMENT_FILEPATH: "/etc/kubernetes/custom_environment.json"
+            ${{ if eq(driverWriteSecret, 'true') }}:
+              DRIVER_WRITE_SECRETS: true
```
Comment (ritazh) [inline .pipelines/templates/e2e-test-kind.yaml:46]: nit: extra space here?

File: pkg/provider/provider.go [hunk 3]
```diff
@@ -283,15 +285,22 @@ func (p *Provider) MountSecretsStoreObjectContent(ctx context.Context, attrib ma
...
-			return nil, errors.Wrapf(err, "failed to write file %s at %s", fileName, targetPath)
+		// if the feature to return secrets to CSI driver isn't enabled, the provider will continue to write
+		// the contents to the filesystem.
+		if !*DriverWriteSecrets {
+			if err := os.WriteFile(filepath.Join(targetPath, fileName), objectContent, permission); err != nil {
+				return nil, nil, errors.Wrapf(err, "failed to write file %s at %s", fileName, targetPath)
+			}
+			klog.InfoS("successfully wrote file", "file", fileName, "pod", klog.ObjectRef{Namespace: p.PodNamespace, Name: p.PodName})
+		} else {
+			// these files will returned to the CSI driver as part of grpc response
```
Comment (ritazh) [inline pkg/provider/provider.go:298]: nit:
```suggestion
			// these files will be returned to the CSI driver as part of grpc response
```

File: pkg/provider/provider.go [hunk 4]
```diff
@@ -283,15 +285,22 @@ func (p *Provider) MountSecretsStoreObjectContent(ctx context.Context, attrib ma
...
+			files[fileName] = objectContent
```
Comment (ritazh) [inline pkg/provider/provider.go:299]: should we still log success here? would be useful when we debug issues between driver and provider.

**Review by ritazh — APPROVED** [review]
LGTM
just few minor nits

File: pkg/provider/provider.go (hunk 4 above)
Comment (aramase) [reply pkg/provider/provider.go:299]: Done!

File: .pipelines/templates/e2e-test-kind.yaml (hunk 2 above)
Comment (aramase) [reply .pipelines/templates/e2e-test-kind.yaml:46]: As it's part of the `if` clause the space is required

**Pull request merged**